File: src/tools/bwrap-sandbox.ts
/**
 * bubblewrap (bwrap) 沙箱集成。
 *
 * 当系统安装了 bwrap 时，Bash 工具的子进程会在隔离沙箱中执行；
 * 未安装时自动回退到直接执行（当前行为）。
 *
 * @author 真心
 */

import { existsSync } from "node:fs";

/**
 * 子进程启动选项。
 */
interface SpawnOptions {
  readonly stdout: "pipe";
  readonly stderr: "pipe";
  readonly stdin: "ignore";
}

/** bwrap 可用性缓存：undefined = 未检测，true/false = 已检测 */
let bwrapAvailable: boolean | undefined;

/** 需要传递到沙箱内的最小环境变量名列表 */
const MINIMAL_ENV_KEYS = ["HOME", "PATH", "TMPDIR", "LANG", "TERM"] as const;

/** 沙箱内以只读方式挂载的系统目录 */
const READONLY_BIND_DIRS = ["/usr", "/bin", "/lib", "/lib64", "/etc"] as const;

/**
 * 检测 bwrap 是否可用。结果会被缓存，整个进程生命周期只检测一次。
 */
export async function isBubblewrapAvailable(): Promise<boolean> {
  if (bwrapAvailable !== undefined) {
    return bwrapAvailable;
  }

  try {
    const proc = Bun.spawn({
      cmd: ["bwrap", "--version"],
      stdin: "ignore",
      stdout: "pipe",
      stderr: "pipe"
    });
    const exitCode = await proc.exited;
    bwrapAvailable = exitCode === 0;
    return bwrapAvailable;
  } catch (_error: unknown) {
    bwrapAvailable = false;
    return false;
  }
}

/**
 * 直接启动子进程（无沙箱），即原始的 Bun.spawn 行为。
 */
export function spawnDirect(
  command: string,
  workspaceRoot: string,
  options: SpawnOptions
): Bun.Subprocess<"ignore", "pipe", "pipe"> {
  return Bun.spawn({
    cmd: ["zsh", "-lc", command],
    cwd: workspaceRoot,
    stdin: options.stdin,
    stdout: options.stdout,
    stderr: options.stderr
  });
}

/**
 * 在 bwrap 沙箱中启动子进程。
 *
 * 沙箱策略：
 * - `--unshare-all` 隔离所有命名空间
 * - 系统目录（/usr, /bin, /lib, /lib64, /etc）只读挂载
 * - 工作区读写挂载
 * - /tmp 使用 tmpfs（沙箱内独立）
 * - /proc 和 /dev 按需挂载以支持常见命令
 * - 仅传递最小环境变量 + BANKA_* 变量
 */
export function spawnSandboxed(
  command: string,
  workspaceRoot: string,
  options: SpawnOptions
): Bun.Subprocess<"ignore", "pipe", "pipe"> {
  const args = buildBwrapArgs(command, workspaceRoot);
  return Bun.spawn({
    cmd: args,
    cwd: workspaceRoot,
    stdin: options.stdin,
    stdout: options.stdout,
    stderr: options.stderr
  });
}

/**
 * 构建 bwrap 命令参数列表。
 */
function buildBwrapArgs(command: string, workspaceRoot: string): string[] {
  const args: string[] = [
    "bwrap",
    "--unshare-all",
    "--new-session",
    "--die-with-parent",
    "--clearenv",
    "--proc",
    "/proc",
    "--dev",
    "/dev"
  ];

  for (const dir of READONLY_BIND_DIRS) {
    if (existsSync(dir)) {
      args.push("--ro-bind", dir, dir);
    }
  }

  args.push("--tmpfs", "/tmp");
  args.push("--bind", workspaceRoot, workspaceRoot);

  const envVars = collectMinimalEnv();
  for (const [key, value] of Object.entries(envVars)) {
    args.push("--setenv", key, value);
  }

  args.push("zsh", "-lc", command);
  return args;
}

/**
 * 收集需要传入沙箱的最小环境变量集合。
 */
function collectMinimalEnv(): Record<string, string> {
  const env: Record<string, string> = {};

  for (const key of MINIMAL_ENV_KEYS) {
    const value = process.env[key];
    if (value !== undefined && value !== "") {
      env[key] = value;
    }
  }

  for (const [key, value] of Object.entries(process.env)) {
    if (key.startsWith("BANKA_") && value !== undefined && value !== "") {
      env[key] = value;
    }
  }

  return env;
}

export type { SpawnOptions };

File: internal/tools/bwrap_sandbox.go
package tools

import (
	"context"
	"os"
	"os/exec"
	"sort"
	"strings"
	"sync"
)

var (
	bwrapCheckOnce sync.Once
	bwrapAvailable bool
)

var minimalEnvironmentKeys = []string{"HOME", "PATH", "TMPDIR", "LANG", "TERM"}
var readonlyBindDirectories = []string{"/usr", "/bin", "/lib", "/lib64", "/etc"}

func isBubblewrapAvailable() bool {
	bwrapCheckOnce.Do(func() {
		path, err := exec.LookPath("bwrap")
		if err != nil {
			return
		}
		bwrapAvailable = exec.Command(path, "--version").Run() == nil
	})
	return bwrapAvailable
}

func newShellCommand(ctx context.Context, command string, workspaceRoot string) *exec.Cmd {
	if isBubblewrapAvailable() {
		return exec.CommandContext(ctx, "bwrap", buildBwrapArgs(command, workspaceRoot)...)
	}

	return newDirectShellCommand(ctx, command)
}

func newDirectShellCommand(ctx context.Context, command string) *exec.Cmd {
	shell := "zsh"
	if _, err := exec.LookPath(shell); err != nil {
		shell = "bash"
	}
	cmd := exec.CommandContext(ctx, shell, "-lc", command)
	cmd.Env = sanitizedChildEnvironment(nil)
	return cmd
}

func sanitizedChildEnvironment(overrides map[string]string) []string {
	values := make(map[string]string)
	for _, entry := range os.Environ() {
		key, value, ok := strings.Cut(entry, "=")
		if ok && !strings.HasPrefix(key, "BANKA_") {
			values[key] = value
		}
	}
	for key, value := range overrides {
		values[key] = value
	}
	keys := make([]string, 0, len(values))
	for key := range values {
		keys = append(keys, key)
	}
	sort.Strings(keys)
	result := make([]string, 0, len(keys))
	for _, key := range keys {
		result = append(result, key+"="+values[key])
	}
	return result
}

func buildBwrapArgs(command string, workspaceRoot string) []string {
	args := []string{
		"--unshare-all",
		"--new-session",
		"--die-with-parent",
		"--clearenv",
		"--proc", "/proc",
		"--dev", "/dev",
	}

	for _, directory := range readonlyBindDirectories {
		if _, err := os.Stat(directory); err == nil {
			args = append(args, "--ro-bind", directory, directory)
		}
	}
	args = append(args, "--tmpfs", "/tmp")
	args = append(args, "--bind", workspaceRoot, workspaceRoot)
	for _, pair := range collectMinimalEnvironment() {
		args = append(args, "--setenv", pair[0], pair[1])
	}
	args = append(args, "zsh", "-lc", command)
	return args
}

func collectMinimalEnvironment() [][2]string {
	var result [][2]string
	seen := make(map[string]bool)
	for _, key := range minimalEnvironmentKeys {
		if value := os.Getenv(key); value != "" {
			result = append(result, [2]string{key, value})
			seen[key] = true
		}
	}
	return result
}
